(feature): Add ActiveTokenEndpoint to WsFederationConfiguration

victorm-hernandez opened the pull request:
Added support to the WsFederationMetadataSerializer to read the SecurityTokenSerivceEndpoint element from WS-Federation metadata and exposed it via the WsFederationConfiguration object as ActiveTokenEndpoint.

Commit:
(feature): Add ActiveTokenEndpoint to WsFederationConfiguration

Added support to the WsFederationMetadataSerializer to read the
SecurityTokenSerivceEndpoint element from WS-Federation metadata and
exposed it via the WsFederationConfiguration object as ActiveTokenEndpoint.

Commit:
(feature): Add WsFederationConfigurationValidator and related test cases.

Commit:
(feature): Optimized WsFedConfigValidator signature validation.

Instead of attempting with each of the signing keys, we will match the
key id of the signature with the key id of the signing key.

brentschmaltz commented on src/Microsoft.IdentityModel.Protocols.WsFederation/WsFederationMetadataSerializer.cs:
```diff
@@ -248,6 +258,75 @@ protected virtual string ReadPassiveRequestorEndpoint(XmlReader reader)
+            reader.MoveToContent();
+            reader.ReadEndElement();
+
+            // </PassiveRequestorEndpoint>
```
Should this comment be: // SecurityTokenServiceEndpoint #Resolved

victorm-hernandez replied:
Yes, updating

TimHannMSFT replied:
Looks like this was marked resolved but also not updated in the branch so far as I can tell. @victorm-hernandez do you have local changes which still will be persisted?

victorm-hernandez replied:
You are right, to avoid any confusion I will mark elements as resolved once I sent another iteration.

brentschmaltz commented on src/Microsoft.IdentityModel.Tokens/BaseConfiguration.cs:
```diff
@@ -27,11 +27,17 @@ public virtual ICollection<SecurityKey> SigningKeys
 
+        /// <summary>
+        /// Gets or sets the token endpoint specified via the metadata endpoint.
+        /// This can is the fed:SecurityTokenServiceType in WS-Federation, http://docs.oasis-open.org/wsfed/federation/v1.2/os/ws-federation-1.2-spec-os.html#:~:text=fed%3ASecurityTokenSerivceEndpoint
```
'This can is' => 'This is' #Resolved

victorm-hernandez replied:
Fixed

TimHannMSFT replied:
Seems like it's not yet fixed?

brentschmaltz commented on src/Microsoft.IdentityModel.Protocols.WsFederation/WsFederationMetadataSerializer.cs:
```diff
@@ -248,6 +258,75 @@ protected virtual string ReadPassiveRequestorEndpoint(XmlReader reader)
+
+            string tokenEndpoint = null;
+
+            while (reader.IsStartElement())
```
If the reader is not positioned on a StartElement what will happen? #Resolved

victorm-hernandez replied:
Since we know that the XML is not malformed, the only scenario where the reader is not on a start element is if the endpointreference has no inner tags and ends right there.  

<wsa:EndpointReference xmlns:wsa=""[http://www.w3.org/2005/08/addressing](http://www.w3.org/2005/08/addressing)""></wsa:EndpointReference>

In that case the MoveToContent below will have no effect and finish reading the SecurityTokenServiceEndpoint tag.

The return value will be null for tokenEndpoint.

victorm-hernandez replied:
The configuration validator class has a check for null on this URI and would catch that problem.

brentschmaltz commented on src/Microsoft.IdentityModel.Protocols.OpenIdConnect/Configuration/OpenIdConnectConfiguration.cs:
```diff
@@ -317,6 +317,12 @@ public OpenIdConnectConfiguration(string json)
+        /// This base class property is not used in OpenIdConnect. 
+        /// </summary>
+        [JsonIgnore]
+        public override string ActiveTokenEndpoint { get; set; }
```
We will modify this in our 7.x release when we will use the utf8Reader and set the ActiveTokenEndpoint to the "token_endpoint" and PassiveTokenEndpoint to "authorization_endpoint".

TimHannMSFT replied:
Are we keeping track of the changes we're intending to make/are in the process of making somewhere? We'll definitely want to make sure we have really clear release notes when we go to 7.x so it's clear who's impacted.

This comment is more an FYI for the community then? There's no action for this PR correct?

brentschmaltz commented on src/Microsoft.IdentityModel.Protocols.WsFederation/WsFederationConstants.cs:
```diff
@@ -5,6 +5,7 @@ namespace Microsoft.IdentityModel.Protocols.WsFederation
 {
     /// <summary>
     /// Constants for WsFederation.
+    /// As defined on the http://docs.oasis-open.org/wsfed/federation/v1.2/os/ws-federation-1.2-spec-os.html
```
'As defined on the' => 'As defined in: ' #Resolved

victorm-hernandez replied:
done

FuPingFranco commented on src/Microsoft.IdentityModel.Protocols.WsFederation/Configuration/WsFederationConfigurationValidator.cs:
```diff
@@ -0,0 +1,195 @@
+        /// Validates a WsFederationConfiguration.
+        /// </summary>
+        /// <param name="configuration">WsFederationConfiguration to validate</param>
+        /// <returns>A <see cref="ConfigurationValidationResult"/> that contains validation result.</returns>
```
nit: "that contains validation result" -> "containing the validation result" #Resolved

FuPingFranco commented on src/Microsoft.IdentityModel.Protocols.WsFederation/WsFederationMetadataSerializer.cs:
```diff
@@ -248,6 +258,75 @@ protected virtual string ReadPassiveRequestorEndpoint(XmlReader reader)
         }
 
+        /// <summary>
+        /// Read fed:SecurityTokenServiceEndpoint element from metadata XML.
```
nit: "Read" -> "Reads"

FuPingFranco reviewed (approved):
Looks great!

sruke commented on src/Microsoft.IdentityModel.Protocols.WsFederation/Configuration/WsFederationConfigurationValidator.cs:
```diff
@@ -0,0 +1,195 @@
+
+            return new ConfigurationValidationResult
+            {
+                ErrorMessage = LogMessages.IDX22711,
```
CryptographicException and signature verification failure log the same message. Consider logging a clearer log message in both places to indicate the reason for signature validation failure.  #Resolved

victorm-hernandez replied:
Good point, added IDX22712

TimHannMSFT commented on src/Microsoft.IdentityModel.Protocols.WsFederation/Configuration/WsFederationConfigurationValidator.cs:
```diff
@@ -0,0 +1,195 @@
+using System.Security.Cryptography.X509Certificates;
+using Microsoft.IdentityModel.Tokens;
+using Microsoft.IdentityModel.Xml;
+using static Microsoft.IdentityModel.Logging.LogHelper;
```
TIL about using static #Pending

victorm-hernandez replied:
Could you clarify?

TimHannMSFT commented on src/Microsoft.IdentityModel.Protocols.WsFederation/Configuration/WsFederationConfigurationValidator.cs:
```diff
@@ -0,0 +1,195 @@
+                    ErrorMessage = LogMessages.IDX22700,
+                    Succeeded = false
+                };
+            }
```
Consider doing all these checks up front and use StringBuilder to build out an error message. In the current model a developer may face one issue, update, face another issue, update...etc. Getting all the errors up front allows for more actionable errors/fewer iterations of debugging. Also less code maintenance costs IMO since if we decide to add another check it doesn't require another log message.  #WontFix

victorm-hernandez replied:
The current code assume some of the checks before it. While is not a bad idea it requires a refactor of the whole method. If you dont mind I would like to postpone this improvement.

TimHannMSFT commented on src/Microsoft.IdentityModel.Protocols.WsFederation/Configuration/WsFederationConfigurationValidator.cs:
```diff
@@ -0,0 +1,195 @@
+
+                try
+                {
+                    configuration.Signature.Verify(key, key.CryptoProviderFactory);
```
Looks like this can throw `ArgumentNullException` too. That would be a bit odd since in other cases where there's unexpected input it's returned as a failed `ConfigurationValidationResult`. Is this intentional? #Resolved

victorm-hernandez replied:
This is a good catch. 

On the our scenario this is not a real problem. If we deserialize the configuration using the metadataSerializer class the security key always use the default crypto provider and this value is never null.

Having said that, it is possible to modify this property before calling the validator. We need to handle that scenario.

I will update the logic to check if there is a crypto provider before attempting to verify.

TimHannMSFT commented on src/Microsoft.IdentityModel.Protocols.WsFederation/LogMessages.cs:
```diff
@@ -23,17 +23,33 @@ internal static class LogMessages
         internal const string IDX22810 = "IDX22810: 'Issuer' value is missing in wsfederationconfiguration.";
         internal const string IDX22811 = "IDX22811: 'TokenEndpoint' value is missing in wsfederationconfiguration.";
         internal const string IDX22812 = "IDX22812: Element: '{0}' was an empty element. 'TokenEndpoint' value is missing in wsfederationconfiguration.";
+        internal const string IDX22813 = "IDX22813: 'ActiveTokenEndpoint' is missing in 'SecurityTokenServiceTypeRoleDescriptor'.";
```
nit: [consistency] this is in single quotes here and SecurityTokenServiceEndpoint is not in the next log. Also IDX22806, IDX22807 talk about "token service type RoleDescriptor". I think the way you have it is more clear probably but good to have consistency too for ease of understanding. #Resolved

victorm-hernandez replied:
Good point, will update.

TimHannMSFT reviewed (approved):
Did not complete looking through all the testing but the core changes LGTM, left a few comments where I believe some improvements could be considered.

dannybtsai commented on src/Microsoft.IdentityModel.Protocols.WsFederation/WsFederationMetadataSerializer.cs:
```diff
@@ -191,6 +195,9 @@ protected virtual SecurityTokenServiceTypeRoleDescriptor ReadSecurityTokenServic
             if (string.IsNullOrEmpty(roleDescriptor.TokenEndpoint))
                 LogHelper.LogWarning(LogMessages.IDX22807);
 
+            if (string.IsNullOrEmpty(roleDescriptor.ActiveTokenEndpoint))
```
This implies that roleDescriptor.ActiveTokenEndpoint can be null or empty (not an error), but ReadSecurityTokenServiceEndpoint method throws LogMessages.IDX22812 exception if the element is empty. #Resolved

victorm-hernandez replied:
This is correct. I understand how this can be confusing. 

This inconsistency is there to avoid a breaking change ActiveTokenEndpoint is a new property that may not be used today by consumers of our library, if they chose to populate this class in any way but our metadata serializer this value may be null. So we cannot really enforce this new property without introducing a breaking change. 

The story is different for the configuration validator. This is a new class that no one is using so it can properly validate that the field exists and has the correct value.

dannybtsai commented on src/Microsoft.IdentityModel.Protocols.WsFederation/WsFederationMetadataSerializer.cs:
```diff
@@ -248,6 +258,75 @@ protected virtual string ReadPassiveRequestorEndpoint(XmlReader reader)
+                    reader.ReadStartElement(WsAddressing.Elements.Address, WsAddressing.Namespace);
+                    reader.MoveToContent();
+
+                    tokenEndpoint = Trim(reader.ReadContentAsString());
```
Do we need to catch exception from ReadContentAsString? #Resolved

victorm-hernandez replied:
There is no need, the entry point or this flow has a try/catch block that handles all exceptions

dannybtsai commented on src/Microsoft.IdentityModel.Protocols.WsFederation/Configuration/WsFederationConfigurationValidator.cs:
```diff
@@ -0,0 +1,195 @@
+                };
+            }
+
+            if (string.IsNullOrWhiteSpace(configuration.Signature.SignedInfo.SignatureMethod))
```
null check SignedInfo #Resolved

victorm-hernandez replied:
good catch

dannybtsai commented on src/Microsoft.IdentityModel.Protocols.WsFederation/Configuration/WsFederationConfigurationValidator.cs:
```diff
@@ -0,0 +1,195 @@
+                };
+            }
+
+            if (configuration.Signature.SignedInfo.References == null || configuration.Signature.SignedInfo.References.Count == 0)
```
null check SignedInfo #Resolved

dannybtsai commented on src/Microsoft.IdentityModel.Protocols.WsFederation/Configuration/WsFederationConfigurationValidator.cs:
```diff
@@ -0,0 +1,195 @@
+                        Succeeded = true
+                    };
+                }
+                catch (XmlValidationException)
```
1. Should catch other types of exceptions too?
2. Should break or continue? #Pending

victorm-hernandez replied:
Although the only exception expected is this, the cryptoprovider may potentially throw any other exception.

We could eat the exception and try to continue however given that we checked for the key thumbprint we do not expect any other key to be valid. We do not expect multiple entries with the same thumbprint and different crypto provider factory.

In my opinion the best thing to do is to let that exception bubble up. 

The other alternative I see would be to update the ConfigurationValidationResult object to include exception information, given that that object is shared across ConfigurationValidators I would prefer to not change the pattern.

RojaEnnam replied:
We do want to try rest of the keys in the list if one fails correct? In that case, like Danny suggested, it should be continue.

victorm-hernandez commented:
thanks for the review

---
In reply to: [1470906811](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/2100#issuecomment-1470906811) [](http://example.com/codeflow?ancestors=1470906811,1470906811)

victorm-hernandez commented:
thanks for the review

---
In reply to: [1472717066](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/2100#issuecomment-1472717066) [](http://example.com/codeflow?ancestors=1472717066,1472717066)

Commit:
(fix): Updated WSFedConfigValidator to handle null properties on the config

New scenarios handled on WsFedConfigValidaot, handle null crypto provider
factory on signing keys, null signing key, null signedinfo.
New error string for WsFedConfigValidator for the scenario where we cannot
find the key used to sign the metadata.
Updated strings for consistency.
Minor XML comment updates.

sruke reviewed (approved):
LGTM!

Commit:
(style/fix): Null check added to WsFedConfigValidator, cosmetic changes

victorm-hernandez commented:
/azp run

RojaEnnam commented on src/Microsoft.IdentityModel.Protocols.WsFederation/SecurityTokenServiceTypeRoleDescriptor.cs:
```diff
@@ -21,13 +21,23 @@ public List<KeyInfo> KeyInfos
 
         /// <summary>
-        /// Token endpoint
+        /// Passive Requestor Token endpoint
+        /// fed:PassiveRequestorEndpoint, https://docs.oasis-open.org/wsfed/federation/v1.2/os/ws-federation-1.2-spec-os.html#:~:text=fed%3ASecurityTokenServiceType/fed%3APassiveRequestorEndpoint
```
Can we add a line or link on what passive requestor is.

RojaEnnam commented on src/Microsoft.IdentityModel.Protocols.WsFederation/SecurityTokenServiceTypeRoleDescriptor.cs:
```diff
@@ -21,13 +21,23 @@ public List<KeyInfo> KeyInfos
         }
+
+        /// <summary>
+        /// Active Requestor Token Endpoint
```
Same as above, can we add some description on what active requestor is.
